framework/preference_buffer.py

```diff
@@ -0,0 +1,45 @@
+from replay_buffer import Trajectory
+import numpy as np
+
+class PreferenceBuffer:
+    def __init__(self, buffer_size):
+        self.buffer_size = buffer_size
+        # Each row contains: first_start_idx, first_end_idx, second_start_idx, second_end_idx, preference
+        self.buffer = np.zeros((self.buffer_size, 5), dtype=np.float32)
+        self.size = 0
+        self.next_idx = 0
+
+    def add(self, first_trajectory: Trajectory, second_trajectory: Trajectory, preference: float):
+        """
+        Adds a new preference to the buffer. If the buffer is full, it overwrites the oldest entry.
+
+        :param first_trajectory: First trajectory being compared
+        :param second_trajectory: Second trajectory being compared
+        :param preference: Preference score (e.g., 1 if first is preferred, -1 if second is preferred, 0 for equal preference)
```

Comment:
Is there a reason that the preference values returned by give_preference (2.0, 0.5, 1.0) are not aligned with the expected values in the add method (-1, 0, 1)?

Reply:
I think using (-1, 0, 1) for preferences as it aligns better with our preference_loss() implementation. This format ensures that the loss calculation (-torch.sum(preferences * torch.log(predictions))) operates as intended, with -1 representing a preference for the second option, 1 for the first option, and 0 for equal preference. Using any other format, such as (2.0, 1.0, 0.5), would require additional preprocessing and could lead to inconsistencies in the loss calculation.

Reply:
I just noticed in the PEBBLE paper they use (0, 0.5, 1) and the loss looks like Binary Cross-Entropy.